### PR TITLE
Redirect unknown users from login to signup page

### PR DESCRIPTION
**Pull request by monsieur-z**
### Changes proposed in this Pull Request

With this PR, we redirect users that submit the login form with an unknown username to the signup form.

Fixes 1199916399796129-as-1199916399796148

### Implementation notes

- The unknown-user redirect behavior is behind the `jetpack/simplify-login-signup-flows` feature flag, which is turned **_on_** for 'development' and 'staging' environments. To temporarily disable the feature, you can add `&flags=-jetpack/simplify-login-signup-flows` to the url.  (notice the minus sign(-) before the feature name)
- Also the redirect only takes place during Jetpack sign-in flow (Jetpack connect flow sign-in or Jetpack Cloud sign-in).  (**Not** WordPress.com, or WooCommerce, or Gutenboarding sign-in)

To-dos:
- memoize `getSignupUrl` (nice-to-have)

### Testing instructions

- Download the PR and run Calypso

##### Jetpack Connect

1. Start up a Jurassic Ninja site and open it in an incognito window.
2. Go through the connect flow until  you come to the log in page.
3. Copy URL, paste it in a new browser tab and replace `https://wordpress.com/` with `http://calypso.localhost:3000/`.
4. Enter a username or email you know is not related to any WordPress account
5. Click 'Continue" button and verify that you're redirected to the new user signup form. (instead of shown a "User doesn't exist" form validation message)
6. In the URL replace `https://wordpress.com/` with `http://calypso.localhost:3000/` again.
7. Verify that the email or username entered in in step 4 is carried over to the signup form.

##### Jetpack Cloud

1. Open cloud.jetpack.com in an incognito window. You should be redirected to `https://wordpress.com/log-in`
2. In the URL replace `https://wordpress.com/` with `http://calypso.localhost:3000/`.
3. Enter a username or email you know is not related to any WordPress account
4. Click 'Continue" button and verify that you're redirected to the new user signup form. (instead of shown a "User doesn't exist" form validation message)
5. Verify that the email or username entered in in step 3 has carried over to the signup form.

##### WordPress.com Signup Flow (Not redirected)

1. Visit `http://calypso.localhost:3000/log-in`.
2. Enter a username or email you know is not related to any WordPress account
3. Verify that you are **not** redirected to sign-up form, but instead are shown the "User doesn't exist" form validation error message.

##### WooCommerce (Not redirected)

1. Visit `woocommerce.com` and click "Log in" in the nav bar. In the URL replace `https://wordpress.com` with `http://calypso.localhost:3000`.
2. Enter a username or email you know is not related to any WordPress account
3. Verify that you are **not** redirected to sign-up form, but instead are shown the "User doesn't exist" form validation error message.

##### Gutenboarding (Not redirected)

1. Visit `calypso.localhost:3000/new`. 
2. Advance through all the prompts, select a plan, and when you come to the "Save your progress" page. Instead of creating a new account, click the 'Already have an account? **Log in**' link.
3. Enter a username or email you know is not related to any WordPress account
3. Verify that you are **not** redirected to sign-up form, but instead are shown the "User doesn't exist" form validation error message.

### Screenshots

Expected behaviour

https://user-images.githubusercontent.com/1620183/107568188-a5825700-6bb4-11eb-9ff6-8060cee12cc6.mov

### CODE REVIEW

File: client/blocks/login/login-form.jsx [hunk 1]
```diff
@@ -105,8 +107,12 @@ export class LoginForm extends Component {
 			! disableAutoFocus && defer( () => this.password && this.password.focus() );
 		}
 
-		if ( requestError.field === 'usernameOrEmail' ) {
+		if ( this.showUsernameError() ) {
 			! disableAutoFocus && defer( () => this.usernameOrEmail && this.usernameOrEmail.focus() );
+		} else {
+			window.location.replace(
```
Comment (monsieur-z) [inline client/blocks/login/login-form.jsx]: `page.redirect` doesn't trigger a render of the signup component (the `start` controller is not called).

File: client/blocks/login/login-form.jsx [hunk 2]
```diff
@@ -158,6 +164,74 @@ export class LoginForm extends Component {
 		return ! socialAccountIsLinking && ! hasAccountTypeLoaded;
 	}
 
+	getSignupUrl() {
```
Comment (monsieur-z) [inline client/blocks/login/login-form.jsx]: Extracted the logic from the `render` function.

File: client/blocks/login/login-form.jsx [hunk 3]
```diff
@@ -158,6 +164,74 @@ export class LoginForm extends Component {
...
+	}
+
+	showUsernameError() {
+		const { requestError } = this.props;
+
+		// TODO: add feature flag or Jetpack context condition
+		return (
+			requestError &&
+			requestError.field === 'usernameOrEmail' &&
+			requestError.code !== LOGIN_ERROR_UNKNOWN_USER
```
Comment (monsieur-z) [inline client/blocks/login/login-form.jsx]: This prevents the login form from showing an error while the signup page is loading. The transition is too fast to read anything.

File: client/blocks/login/login-form.jsx [hunk 4]
```diff
@@ -408,64 +482,15 @@ export class LoginForm extends Component {
...
 			isJetpackWooDnaFlow,
 			wccomFrom,
-			currentRoute,
-			currentQuery,
-			pathname,
-			locale,
 		} = this.props;
 		const isOauthLogin = !! oauth2Client;
 		const isPasswordHidden = this.isUsernameOrEmailView();
-
```
Comment (monsieur-z) [inline client/blocks/login/login-form.jsx]: Extracted this bit in `getSignupUrl`.

**Review by mattgawarecki — APPROVED** [review]
Testing worked for me, and I don't see any issues with the code. I appreciate how you separated the logic in `showUsernameError` and `getSignupUrl`!

**Comment by monsieur-z** [comment]
I noticed I could use `page` instead of `window.location.replace` to prevent a browser reload.

**Comment by keoshi** [comment]
This is really exciting!

File: client/blocks/login/login-form.jsx [hunk 5]
```diff
@@ -140,6 +145,20 @@ export class LoginForm extends Component {
...
 	};
 
+	isJetpackConnectLogin() {
+		const { currentRoute } = this.props;
+		return startsWith( currentRoute, '/log-in/jetpack' );
+	}
+
+	isJetpackCloudLogin() {
+		const { currentRoute, currentQuery } = this.props;
+		return startsWith( currentRoute, '/log-in' ) && currentQuery?.client_id;
```
Comment (rcanepa) [inline client/blocks/login/login-form.jsx]: Can we somehow check that `client_id` is indeed a Jetpack cloud client ID? It concerns me a bit that there are other clients (Woo comes to mind) that this code might consider as if they were a Jetpack cloud client id.

**Comment by rcanepa** [comment]
Looks great overall! Left a comment about checking the `client_id`. 

One thing I noticed, is that when I came from the setup process of a JN site, the email I entered was lost when I was redirected from login to signup. I guess there is something else we need to do there in order to preserve the email entered by the user in the login form. Here is what I saw:

LoginSignup

File: client/blocks/login/login-form.jsx (hunk 5 above)
Comment (elliottprogrammer) [reply client/blocks/login/login-form.jsx]: Yes, very good point!  Ok yes, we can utilize the `getCurrentOAuth2Client()` selector to check that it is a `jetpack-cloud` client_id. 👍

**Comment by elliottprogrammer** [comment]
@rcanepa, Ok I improved `isJetpackCloud()` to check if the `client_id` is a "jetpack-cloud" client_id.
Also when entering an unknown username or email, the username or email will carry over to the sign up form. 👍

**Comment by rcanepa** [comment]
> @rcanepa, Ok I improved `isJetpackCloud()` to check if the `client_id` is a "jetpack-cloud" client_id.

Nice!

> Also when entering an unknown username or email, the username or email will carry over to the sign up form. 👍

This is a really nice improvement.

**Comment by rcanepa** [comment]
I think this PR is mostly done. The only thing that comes to mind is that in #50008, @robertf4 abstracted the signup URL construction into a utility function that we should use here (to avoid duplication). Other than that, I think the PR is finished and we can start to think about how we can show this to design so they can weigh in and decide if we want to add this improvement or no.

**Comment by elliottprogrammer** [comment]
> I think this PR is mostly done. The only thing that comes to mind is that in #50008, @robertf4 abstracted the signup URL construction into a utility function that we should use here (to avoid duplication). Other than that, I think the PR is finished and we can start to think about how we can show this to design so they can weigh in and decide if we want to add this improvement or no.

@rcanepa, I integrated the `getSignupUrl()` function that Robert abstracted, and carrying over the username or email to the signup form should be all good now for Jetpack cloud login and Jetpack connect.  👍  However it appears canary e2e tests are failing..  Looking into it now..

**Comment by elliottprogrammer** [comment]
Closing this PR because the new signup flow ([magic links](https://github.com/Automattic/wp-calypso/pull/50203)) makes this PR unnecessary.
p1HpG7-bae-p2

**Pull request closed without merging**